File: src/main/java/com/github/sheauoian/croissantsys/command/CroissantCMD.java
package com.github.sheauoian.croissantsys.command;

import com.github.sheauoian.croissantsys.CroissantSys;
import com.github.sheauoian.croissantsys.common.user.OnlineUser;
import com.github.sheauoian.croissantsys.common.user.UserManager;
import org.bukkit.Location;
import org.bukkit.command.Command;
import org.bukkit.command.CommandSender;
import org.bukkit.command.TabCompleter;
import org.bukkit.entity.Player;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.Arrays;
import java.util.List;

public class CroissantCMD extends CMD implements TabCompleter {
    private final String[] completeList = new String[] {"help", "heal", "setspawn", "menu"};
    public CroissantCMD(CroissantSys plugin) {
        super(plugin);
    }

    @Override
    public boolean onCommand(
            @NotNull CommandSender sender,
            @NotNull Command cmd,
            @NotNull String text,
            @NotNull String[] args)
    {
        if (args.length >= 1) {
            switch (args[0]) {
                case "help" -> {
                    sender.sendMessage("/sleep version : バージョンを表示します");
                    sender.sendMessage("/sleep help : ヘルプを表示します");
                }
                case "heal" -> {
                    if (sender instanceof Player p) {
                        UserManager.getInstance().getOnlineUser(p).resetHealth();
                        p.sendMessage("HPをリセットしました");
                    } else {
                        sender.sendMessage("このコマンドはプレイヤーのみぞつかうことができよう");
                    }
                }
                case "setspawn" -> {
                    if (sender instanceof Player p) {
                        CroissantSys plugin = CroissantSys.getInstance();
                        Location spawn = p.getWorld().getBlockAt(p.getLocation()).getLocation();
                        spawn.setYaw(p.getYaw());
                        plugin.getConfig().set("spawnpoint", spawn);
                        plugin.saveConfig();
                        p.sendMessage("ようこそ");
                    } else {
                        sender.sendMessage("このコマンドはプレイヤーのみぞつかうことができよう");
                    }
                }
                case "menu" -> {
                    if (sender instanceof Player p) {
                        OnlineUser user = UserManager.getInstance().getOnlineUser(p);
                        user.getMenu().open();
                    }
                }
            }
        }
        return false;
    }

    @Override
    public @Nullable List<String> onTabComplete(
            @NotNull CommandSender commandSender,
            @NotNull Command command,
            @NotNull String s,
            @NotNull String[] args)
    {
        if (args.length == 1) {
            return Arrays.asList(this.completeList);
        }
        return null;
    }

    @Override
    CMD getInstance() {
        return this;
    }
    @Override
    public String getCommandName() {
        return "croissant";
    }
}

File: src/main/java/com/github/sheauoian/croissantsys/command/CroissantCmd.kt
package com.github.sheauoian.croissantsys.command

import com.github.sheauoian.croissantsys.CroissantSys.Companion.instance
import org.bukkit.command.Command
import org.bukkit.command.CommandSender
import org.bukkit.command.TabCompleter
import org.bukkit.entity.Player

class CroissantCmd : TabCompleter, Cmd() {
    override fun onCommand(sender: CommandSender, cmd: Command, text: String, args: Array<String>): Boolean {
        if (args.isNotEmpty()) {
            when (args[0]) {
                "initialspawn" -> { // 初期スポーン地点の設定を行います。
                    if (args.size >= 2) {
                        when (args[1]) {
                            "set" -> {
                                if (sender is Player) {
                                    instance.let {
                                        it.initialSpawnPoint = sender.location
                                        sender.sendMessage("初期スポーン地点を設定しました。")
                                    }
                                }
                            }
                            "tp" -> {
                                if (sender is Player) {
                                    instance.initialSpawnPoint.let { sender.teleport(it) }
                                    sender.sendMessage("初期スポーン地点にテレポートしました。")
                                }
                            }
                        }
                    } else {
                        if (sender is Player) {
                            instance.initialSpawnPoint.let { sender.teleport(it) }
                            sender.sendMessage("初期スポーン地点にテレポートしました。")
                        } else {
                            sender.sendMessage(instance.initialSpawnPoint.toString())
                        }
                    }
                }

                "reload" -> {

                }
            }
        }
        return false
    }

    override fun onTabComplete(sender: CommandSender, cmd: Command, s: String, args: Array<String>): List<String>? {
        if (args.size == 1) {
            return listOf("initialspawn", "reload")
        } else if (args.size == 2) {
            when (args[1]) {
                "initialspawn" -> {
                    return listOf("set", "tp")
                }
            }
        }
        return null
    }

    override val commandName = "croissant"
}
